File: 2022/rust/src/day2.rs
#[allow(dead_code)]
use crate::utils;

pub fn day_2(data: String) {
  println!("Day 1:");
  println!("- Part 1: {}", part_1(&data));
  // println!("- Part 2: {}", day_1_part_2(&data));
}

static RPS: [[u32; 3]; 3] = [
  [3, 6, 0],
  [0, 3, 6],
  [6, 0, 3],
];

fn values_to_index(values: &str) -> [u32; 2] {
  return utils::vec_to_n_array::<u32, 2>(
    values
      .split(" ")
      .map(|value| {
        match value {
          "A" | "X" => 0,
          "B" | "Y" => 1,
          "C" | "Z" => 2,
          _ => panic!("Wrong input value!"),
        }
      })
      .collect::<Vec<u32>>()
  );
}

fn part_1(data: &String) -> u32 {
  return data
    .split("\n")
    .map(|values| {
      let [f, s] = values_to_index(values);
      return RPS[f as usize][s as usize] + s + 1;
    })
    .sum::<u32>();
}

File: archive/2022/src/days/day-2/day2.ts
import { sumArray } from "../../tools";
import { Day } from "../../utils";

const dataToIndex = (data: string): [number, number] => {
  return data.split(" ").map((shape) => {
    switch (shape) {
      case "A":
      case "X":
        return 0;
      case "B":
      case "Y":
        return 1;
      case "C":
      case "Z":
        return 2;
    }
  }) as [number, number];
};

// R P S
const rps = [
  [3, 6, 0], // R
  [0, 3, 6], // P
  [6, 0, 3], // S
];

export default new Day(__filename)
  .addPart(function () {
    return this.input.reduce((sum, data) => {
      const [f, s] = dataToIndex(data);
      return sum + rps[f][s] + s + 1;
    }, 0);
  })
  .addPart(function () {
    return this.input.reduce((sum, data) => {
      const [f, s] = dataToIndex(data);
      if (s === 0) return sum + rps[f].findIndex((i) => i === 0) + 1;
      if (s === 2) return sum + rps[f].findIndex((i) => i === 6) + 6 + 1;
      return sum + 3 + f + 1;
    }, 0);
  });
